src/accumulator/proof.rs

```diff
@@ -0,0 +1,308 @@
+    ///   //Fill `proof_hases` up with all hashes
+    ///   Proof::new(targets, proof_hashes);
+    /// ```  
+    pub fn new(targets: Vec<u64>, hashes: Vec<sha256::Hash>) -> Self {
+        Proof {
+            targets: targets,
+            hashes: hashes,
+        }
+    }
+    /// Public interface for verifying proofs. Returns a result with a bool or an Error
+    /// True means the proof is true given the current stump, false means the proof is
+    /// not valid giving the current stump.
+    ///# Examples
+    /// ```
+    ///   use bitcoin_hashes::{sha256, Hash, HashEngine};
+    ///   use std::str::FromStr;
+    ///   use rustreexo::accumulator::{stump::Stump, proof::Proof};
+    ///   let mut s = Stump::new();
+    ///   // Creates a tree with those values as leafs
+    ///   let test_values:Vec<u8> = vec![0, 1, 2, 3, 4, 5, 6, 7];
+    ///   // Targets are nodes witch we intend to prove
+    ///   let targets = vec![0];
+    ///   
+    ///   let mut proof_hashes = Vec::new();
+    ///   // This tree will look like this
+    ///   // 14                                                         
+    ///   // |-----------------\                               
+    ///   // 12                13                         
+    ///   // |---------\       |--------\               
+    ///   // 08       09       10       11         
+    ///   // |----\   |----\   |----\   |----\       
+    ///   // 00   01  02   03  04   05  06   07
+    ///   // For proving 0, we need 01, 09 and 13's hashes. 00, 08, 12 and 14 can be calculated
+    ///   proof_hashes.push(bitcoin_hashes::sha256::Hash::from_str("4bf5122f344554c53bde2ebb8cd2b7e3d1600ad631c385a5d7cce23c7785459a").unwrap());
+    ///   proof_hashes.push(bitcoin_hashes::sha256::Hash::from_str("9576f4ade6e9bc3a6458b506ce3e4e890df29cb14cb5d3d887672aef55647a2b").unwrap());
+    ///   proof_hashes.push(bitcoin_hashes::sha256::Hash::from_str("29590a14c1b09384b94a2c0e94bf821ca75b62eacebc47893397ca88e3bbcbd7").unwrap());
+    ///   
+    ///   let mut hashes = Vec::new();
+    ///   for i in test_values {
+    ///     let mut engine = bitcoin_hashes::sha256::Hash::engine();
+    ///      engine.input(&[i]);
+    ///      let hash = sha256::Hash::from_engine(engine);
+    ///     hashes.push(hash);
+    ///   }
+    ///
+    ///   s.modify(&hashes, &vec![]);
+    ///   let p = Proof::new(targets, proof_hashes);
+    ///   assert!(p.verify(&vec![hashes[0]] , &s).expect("No error should happens"));
+    ///```
+    pub fn verify(&self, del_hashes: &Vec<sha256::Hash>, stump: &Stump) -> Result<bool, String> {
+        let calculated_roots = self.create_root_candidates(del_hashes, stump)?;
+        let stump_roots = util::num_roots(stump.leafs);
+        let mut number_matched_roots = 0;
+
+        for i in 0..stump_roots {
+            if calculated_roots[number_matched_roots] == stump.roots[stump.roots.len() - (i + 1)] {
```

Comment:
There's an edge case here where if the caller passes in an empty proof, there's an out of bounds error.

```
    { 
      "leafs":[
        0, 1, 2, 3, 4, 5
      ],
      "proof": [
      ],
      "targets": [
      ],
      "expected": true
    }, 
```

It is possible to have an empty proof during ibd if there's an empty block. Could also be an attack vector.